flink-ml-lib/src/main/java/org/apache/flink/ml/feature/standardscaler/StandardScaler.java

```diff
@@ -0,0 +1,288 @@
+            implements OneInputStreamOperator<Row, Tuple3<DenseVector, DenseVector, Long>>,
+                    BoundedOneInput {
+        private ListState<DenseVector> sumState;
+        private ListState<DenseVector> squaredSumState;
+        private ListState<Long> numElementsState;
+        private DenseVector sum;
+        private DenseVector squaredSum;
+        private long numElements;
+
+        private final String featuresCol;
+
+        public ComputeMetaOperator(String featuresCol) {
+            this.featuresCol = featuresCol;
+        }
+
+        @Override
+        public void endInput() {
+            if (numElements > 0) {
+                output.collect(new StreamRecord<>(Tuple3.of(sum, squaredSum, numElements)));
+            }
+        }
+
+        @Override
+        public void processElement(StreamRecord<Row> element) throws Exception {
+            Vector feature = (Vector) element.getValue().getField(featuresCol);
+            if (sum == null) {
+                sum = new DenseVector(feature.size());
+                squaredSum = new DenseVector(feature.size());
+            }
+            BLAS.axpy(1, feature, sum);
+            BLAS.hDot(feature, feature);
+            BLAS.axpy(1, feature, squaredSum);
+            numElements++;
+        }
+
+        @Override
+        public void initializeState(StateInitializationContext context) throws Exception {
+            super.initializeState(context);
+            sumState =
+                    context.getOperatorStateStore()
+                            .getListState(
+                                    new ListStateDescriptor<>(
+                                            "sumState", TypeInformation.of(DenseVector.class)));
+            squaredSumState =
+                    context.getOperatorStateStore()
+                            .getListState(
+                                    new ListStateDescriptor<>(
+                                            "squaredSumState",
+                                            TypeInformation.of(DenseVector.class)));
+            numElementsState =
+                    context.getOperatorStateStore()
+                            .getListState(
+                                    new ListStateDescriptor<>(
+                                            "numElementsState", BasicTypeInfo.LONG_TYPE_INFO));
+
+            sum = OperatorStateUtils.getUniqueElement(sumState, "sumState").orElse(null);
+            squaredSum =
+                    OperatorStateUtils.getUniqueElement(squaredSumState, "squaredSumState")
+                            .orElse(null);
+            numElements =
+                    OperatorStateUtils.getUniqueElement(numElementsState, "numElementsState")
+                            .orElse(0L);
+        }
+
+        @Override
+        public void snapshotState(StateSnapshotContext context) throws Exception {
+            super.snapshotState(context);
+            if (numElements > 0) {
```

Comment:
Good point. 

I agree with the behavior that when the train data is empty, we should output a `special` vector to indicate that we cannot do inference. I have updated the code to handle this case and also added one test case. (See for StandardScalerTest#testFitOnEmptyData).

For other checks about when numElements is zero or not, we probably need to keep it. 

For example, when the training data is not empty, there could be one worker that contains zero data points. If we remove the check, it throws an nullpointer exception.

Reply:
After thinking about this more, I think it might be better to let `estimator.fit(...)` if input data is empty during runtime, so that the error could be moved closer to where it happens, i.e. in the operator that does training instead of the operator that does inference. And this could also make the logic in the model simpler and consistent with other existing models.

What do you think?

Reply:
Thanks for pointint out that "The error should be moved closer to where it happens" and I also believe it is reasonable to throw exception when the training set is empty.

I have addressed this in the latest commit.